File: java/src/calculator/CalculatorConsole.java
package calculator;

import java.io.IOException;

import rspt.runtime.Ref;
import rspt.runtime.Ref_int;

public class CalculatorConsole {
	
	public static void main(String[] args) {
		try {
			System.out.println("CalculatorConsole -- An expression evaluator to showcase RSPT (the Really Simple Parser Tool)");
			System.out.println("(Enter expressions to calculate or 'exit' to quit)");
			CalculatorParser parser = new CalculatorParser();	
			while(true) {
				String      input  = readLine();
				Ref<String> result = new Ref<String>(null);
				Ref_int     error  = new Ref_int(0);
				if(input.equals("exit")) {
					break;
				}
				if(!parser.Parse_ROOT(input.toCharArray(), result, error)) {
		            String next = input.length()-error.val <= 50 ? input.substring(error.val, input.length()) : input.substring(error.val, error.val + 50) + "...";
					String msg = String.format("Error at offset %d: Cannot handle '%s'.", error.val+1, next);
					System.out.println(msg);
				} else {
					String msg = "Result: " + result.val;
					System.out.println(msg);
				}
			}
		} catch(IOException | RuntimeException e) {
			e.printStackTrace();
		}
	}
	
	private static String readLine() throws IOException {
		StringBuilder sb = new StringBuilder();
		while(true) {
			int b = System.in.read();
			if(b == -1 || b == '\n')
				break;
			if(b == '\r')
				continue;
			sb.append((char) b);
		}
		return sb.toString();
	}
}

File: cpp/CalculatorConsole/CalculatorConsole.cpp
#include "stdafx.h"
#include "CalculatorConsole.h"
#include "Parser.h"

#ifdef _DEBUG
#define new DEBUG_NEW
#endif

CWinApp theApp;

int _tmain(int argc, TCHAR* argv[], TCHAR* envp[])
{
	if(argc <= 1) {
		_tprintf(_T("CalculatorConsole -- An expression evaluator to showcase RSPT (the Really Simple Parser Tool)\n"));
		_tprintf(_T("Syntax: $ CalculatorConsole <expression_1> [<expression_2> ... <expression_n>]\n"));
		return 2;
	}

	Parsers::CCalculatorParser p;

	for(int i = 1; i < argc; i++) {
		CString input = argv[i];
		CString result;
		int     error = 0;
		if(!p.Parse_ROOT(input, input.GetLength(), result, error)) {
            CString next = input.GetLength()-error <= 50 ? input.Mid(error, input.GetLength()-error) : input.Mid(error, 50) + "...";
			CString msg; msg.Format(_T("Error at offset %i: Cannot handle '%s'."), error+1, next);
			_tprintf(_T("Error: %s\n"), msg);
		} else {
			_tprintf(_T("Result: %s\n"), result);
		}
	}

	return 0;
}
